### PR TITLE
Networking

### PR DESCRIPTION
**Pull request by Stiffly**
### Edit:

Network logic not using threads anymore
Renamed Package -> Packet
Renamed PopFrontData() -> ReadData(). AddData() -> WriteData().
Renamed PrivateMemberFunctions() -> privateMemberFunctions()
Utilized configfile for address, port, player's name, chosing between server and client.
Deleted NetworkDefinitions.h and restructured its members.
Fixed lines in .ini files.

<b>PlayerSystem</b>
PlayerSystem does not read any input anymore, now updates movement. (with temporary implemented player component)
<b>Rectangle class</b>
Engines Rectangle class conflicted with Asio's rectangle class in Game.cpp for instance. Now using Rectangle::Rectangle(...) to refer to the correct implementation.
<b>Package class</b>
Implemented a Package class to abstract the message passing process with:
### Package.h

``` h
class Package {
    AddPrimitive<type>() { ... } // Adds a primitive type to a package
    AddString();
    AddData();
    PopFrontPrimitive<type>() { ... } // Reads a primitive type from a package
    PopFrontString();
    PopFrontData();
}
```
### Example:

``` cpp
Package package(MessageType::Connect, m_PacketID);
package.AddPrimitive<int>(5); // Player ID
int playerID = package.PopFrontPrimitive<int>(); // playerID will get value of 5
```

<b>Logic:</b>
The network runs on a different thread than the rest of game but has an update function where critical logic can be placed as this will run on the game thread .
Client sends input to server.
Server does movement logic and return transform component.
Client updates its own movement but is forced to the transform component received from the server.

<b>How to use:</b>
Make sure your Networking variable in Config.ini is set to true.
In console write 'C' and press enter for Client and 'S' for Server.
When a server is running and you want to connect with a Client, press 'C' in game.

### CODE REVIEW

File: include/Engine/Network/NetworkDefinitions.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#ifndef NetworkDefines_h__
+#define NetworkDefines_h__
+
+#include <boost/asio/ip/udp.hpp>
+#include <queue>
+#include "Network/Package.h"
+
+
+#define BOARDSIZE 16
```
Comment (sippeangelo) [inline include/Engine/Network/NetworkDefinitions.h]: wtf is this

File: include/Engine/Network/NetworkDefinitions.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+#define MAXCONNECTIONS 8
+#define INPUTSIZE 128
+#define PLAYERSPEED 0.2f;
```
Comment (sippeangelo) [inline include/Engine/Network/NetworkDefinitions.h]: wtf is this

File: include/Engine/Network/Package.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+#include "Network/MessageType.h"
+#include "Core/Util/Logging.h"
+
+// Defines the 
+class Package
+{
+public:
+    // arg1: Type of message (Connect, Disconnect...)
+    // arg2: PackageID for identifying packet loss.
+    Package(MessageType type, unsigned int& packageID);
```
Comment (sippeangelo) [inline include/Engine/Network/Package.h]: Packet != Package

File: src/Engine/Network/Package.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
+#include "Network/Package.h"
+
+Package::Package(MessageType type, unsigned int& packageID)
+{
+    m_Data = new char[128];
+    // Create message header
+    // Add message type
+    int messageType = static_cast<int>(type);
+    Package::AddPrimitive<int>(messageType);
+    packageID = packageID % 1000; // Packet id modulos
```
Comment (sippeangelo) [inline src/Engine/Network/Package.cpp]: ?

File: src/Engine/Network/Package.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
+#include "Network/Package.h"
+
+Package::Package(MessageType type, unsigned int& packageID)
+{
+    m_Data = new char[128];
```
Comment (sippeangelo) [inline src/Engine/Network/Package.cpp]: Template this up and base size on message type?

File: src/Engine/Network/Package.cpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+Package::~Package()
+{
+    delete[] m_Data;
+}
+
+void Package::AddString(std::string str)
+{
+    // Message, add one extra byte for null terminator
+    memcpy(m_Data + m_Offset, str.data(), (str.size() + 1) * sizeof(char));
```
Comment (sippeangelo) [inline src/Engine/Network/Package.cpp]: Missing null terminator?

File: src/Engine/Network/Package.cpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    }
+    memcpy(m_Data + m_Offset, data, sizeOfData);
+    m_Offset += sizeOfData;
+}
+
+std::string Package::PopFrontString()
+{
+    std::string returnValue(m_Data + m_ReturnDataOffset);
+    if (m_Offset < m_ReturnDataOffset + returnValue.size()){
+        LOG_WARNING("Package PopFrontString(): Oh no! You are trying to remove things outside my memory kingdom");
```
Comment (sippeangelo) [inline src/Engine/Network/Package.cpp]: lol

File: resources/DefaultConfig.ini [hunk 8]
```diff
@@ -1,11 +1,19 @@
 [Debug]
+
```
Comment (sippeangelo) [inline resources/DefaultConfig.ini]: Fix line endings

File: src/Engine/Input/InputProxy.cpp [hunk 9]
```diff
@@ -62,7 +62,7 @@ void InputProxy::Process()
             e.Command = command;
             e.Value = currentValue;
             m_EventBroker->Publish(e);
-            LOG_DEBUG("Input: Published command %s=%f for player %i", e.Command.c_str(), e.Value, e.PlayerID);
+            //LOG_DEBUG("Input: Published command %s=%f for player %i", e.Command.c_str(), e.Value, e.PlayerID);
```
Comment (sippeangelo) [inline src/Engine/Input/InputProxy.cpp]: Revert

File: src/Engine/Input/InputProxy.cpp [hunk 10]
```diff
@@ -78,7 +78,7 @@ void InputProxy::Process()
         }
         //e.Value = std::max(-1.f, std::min(e.Value, 1.f));
         m_EventBroker->Publish(e);
-        LOG_DEBUG("Input: Published command %s=%f for player %i", e.Command.c_str(), e.Value, e.PlayerID);
+        //LOG_DEBUG("Input: Published command %s=%f for player %i", e.Command.c_str(), e.Value, e.PlayerID);
```
Comment (sippeangelo) [inline src/Engine/Input/InputProxy.cpp]: Revert

File: src/Engine/Network/Client.cpp [hunk 11]
```diff
@@ -1,11 +1,347 @@
...
+#include "Network/Client.h"
+
+using namespace boost::asio::ip;
 
-Client::Client()
-{
 
+Client::Client() : m_Socket(m_IOService)
+{
+    m_ReceiverEndpoint = udp::endpoint(boost::asio::ip::address::from_string("192.168.1.2"), 13);
```
Comment (sippeangelo) [inline src/Engine/Network/Client.cpp]: *cough*

File: src/Engine/Network/Client.cpp [hunk 12]
```diff
@@ -1,11 +1,347 @@
...
+    // Set up network stream
+    m_NextSnapshot.InputForward = "";
+    m_NextSnapshot.InputRight = "";
 }
 
 Client::~Client()
 {
+    // will it work on linux? #if defined (_WIN64) | defined(_WIN32) otherwise.
+    _CrtDumpMemoryLeaks();
```
Comment (sippeangelo) [inline src/Engine/Network/Client.cpp]: NO

File: src/Game/Game.cpp [hunk 13]
```diff
@@ -53,7 +53,9 @@ Game::Game(int argc, char* argv[])
     m_SystemPipeline->AddSystem<RaptorCopterSystem>();
     m_SystemPipeline->AddSystem<PlayerSystem>();
     m_SystemPipeline->AddSystem<EditorSystem>(m_Renderer);
-
+    // Invoke network
+    if (m_Config->Get<bool>("Networking.StartNetwork", false) == true)
```
Comment (sippeangelo) [inline src/Game/Game.cpp]: https://github.com/teamfisk/TacticalZ/blob/master/tools/CodeRules.h

File: src/Engine/Network/Package.cpp (hunk 6 above)
Comment (Stiffly) [inline src/Engine/Network/Package.cpp]: Probably not

**Pull request merged**